### ISSUE TITLE
Javascript version of the Library

### ISSUE THREAD

**Issue by tqchen** [dmlc/xgboost#1454]
I am opening this issue to see if anyone is interested in contributing.

As XGBoost have minimum dependecy, it is possible to use amalgamation and emscripten to build a javascript version of the library. See https://github.com/dmlc/mxnet.js as an example for our deep learning project.

This could be fun for some usecases, to run XGBoost on the browser and provide some cool demos

**Comment by Jabher** [comment]
@tqchen I can try to participate in it

**Comment by tqchen** [comment]
sounds good, @Jabher Let me know if you need any help on this, mxnet.js could be a good starting point

**Comment by Jabher** [comment]
Cool, thanks!

**Comment by AndrewHannigan** [comment]
@tqchen I personally think that implementing `predict` is more important than implementing `train` in this case.  I can only think of toy examples where a `train` javascript function would be helpful, such as web ML demos.  However, there are a lot of cases where computing xgboost predictions clientside on a pre-built xgboost model would be very practical for a variety of applications.  I've had to do this for clients before using other ML models.  A typical workflow might be:

1. train locally on a research machine using an existing xgboost API (Scala, R, python..).
2. export tree as json object with `xgb.dump`.  
3. read in json model into javascript environment and generate predictions.

The advantage of just implementing the `predict` functionality is that it can be done with VinallaJS.  I could also develop the `predict` functionality quite quickly, whereas a complete javascript API for XGBoost would be a larger-scale undertaking.  What do you think?  Happy to contribute a PR for this.

**Comment by pommedeterresautee** [comment]
@AndrewHannigan FYI there is a project to build compiler of boosting model (transform the model in if else instructions in C++ then compile it). It will be released soon. What you are speaking may be part of this project.
More info here : https://github.com/dmlc/xgboost/issues/2551

**Comment by AndrewHannigan** [comment]
@pommedeterresautee sounds very intriguing, but why take C++ source code as input for the compiler?  Wouldn't JSON be easier to parse, more portable, widely supported by other languages, etc.?  Seems to me the forest should be represented as data, not code.

**Comment by pommedeterresautee** [comment]
performance !!

**Comment by AndrewHannigan** [comment]
Oh I see, so this just going to produce a binary executable.  Sounds cool!  There might be pieces of the pipeline before the ifelse generation step that would overlap with this project.

**Comment by pommedeterresautee** [comment]
@hcho3 may give you more info. 
The parsing of the model and the internal representation may be useful for both projects, plus tree-lite manages also models from lightgbm...

**Comment by hcho3** [comment]
@AndrewHannigan treelite is currently in [public beta](https://github.com/dmlc/treelite). I am adding more documentation before an official release. The idea here is to produce a binary executable for optimized prediction. Right now, treelite produces C program internally to be fed into a C compiler (e.g. gcc), but we can certainly produce a JS program instead.

**Comment by vatsan** [comment]
Has there been any progress on this?

**Comment by JeffersonH44** [comment]
Hello @tqchen,

I have a version on JS working (using WebAssembly) so it's able to run on the browser, also I'm able to load models saved on other programming languages (using th e C API)

[mljs/xgboost](https://github.com/mljs/xgboost)

tell me if you are interested to include it here.

**Comment by tqchen** [comment]
nice, if you are interested in porting back the js binding, we can put it under xgboost/web

We cannot host the emscriptened js file directly in the repo(as it can be big), but we can host the build flow and put emcc generated file in a separate repo.

**Comment by tqchen** [comment]
cc @hcho3

**Comment by tqchen** [comment]
For most of the C++ change, we can use EMSCRIPTEN to detect if the current project is compiled with emcc

**Comment by hcho3** [comment]
@tqchen I'd suppose we need to update both Makefile and CMakeLists.txt to accommodate the JS build?
